File: resolve.go
package skillgetmanager

import (
	"context"
	"fmt"
	"net/url"
	"sort"
	"strings"

	"golang.org/x/mod/semver"
)

// NameVersion holds a parsed skill spec "name" or "name@version".
type NameVersion struct {
	Name    string
	Version string // empty means "latest"
}

// ParseNameVersion splits spec at the last '@' (same rule as the TS prototype).
func ParseNameVersion(spec string) NameVersion {
	at := strings.LastIndex(spec, "@")
	if at <= 0 {
		return NameVersion{Name: spec}
	}
	return NameVersion{
		Name:    spec[:at],
		Version: spec[at+1:],
	}
}

// ResolveInstallTarget resolves a concrete version and archive metadata (no download).
func ResolveInstallTarget(ctx context.Context, spec string) (*VersionDetail, error) {
	nv := ParseNameVersion(spec)
	version := nv.Version
	if version == "" {
		var detail SkillDetail
		path := "/skills/" + url.PathEscape(nv.Name)
		if err := FetchJSON(ctx, path, &detail); err != nil {
			return nil, err
		}
		version = pickInstallableVersion(detail)
		if version == "" {
			return nil, fmt.Errorf("no installable versions for skill %q", nv.Name)
		}
	}
	var vd VersionDetail
	vpath := "/skills/" + url.PathEscape(nv.Name) + "/versions/" + url.PathEscape(version)
	if err := FetchJSON(ctx, vpath, &vd); err != nil {
		return nil, err
	}
	return &vd, nil
}

// pickInstallableVersion chooses the highest semver among non-yanked versions in the
// skill detail map, matching reference registry logic (see getskillpack/registry filestore).
func pickInstallableVersion(d SkillDetail) string {
	if len(d.Versions) == 0 {
		return ""
	}
	var keys []string
	for v, info := range d.Versions {
		v = strings.TrimSpace(v)
		if v == "" || info.Yanked {
			continue
		}
		keys = append(keys, v)
	}
	if len(keys) == 0 {
		return ""
	}
	sort.Slice(keys, func(i, j int) bool {
		return semver.Compare(canonicalSemver(keys[i]), canonicalSemver(keys[j])) < 0
	})
	return keys[len(keys)-1]
}

func canonicalSemver(v string) string {
	if !strings.HasPrefix(v, "v") {
		return "v" + v
	}
	return v
}

File: src/resolve.ts
import { fetchJson } from "./client.js";
import type { SkillDetail, VersionDetail } from "./types.js";

export function parseNameVersion(spec: string): { name: string; version?: string } {
  const at = spec.lastIndexOf("@");
  if (at <= 0) return { name: spec };
  return { name: spec.slice(0, at), version: spec.slice(at + 1) };
}

/** Resolve concrete version and fetch archive metadata (does not download bytes). */
export async function resolveInstallTarget(spec: string): Promise<VersionDetail> {
  const { name, version: pinned } = parseNameVersion(spec);
  let version = pinned;
  if (!version) {
    const detail = await fetchJson<SkillDetail>(`/skills/${encodeURIComponent(name)}`);
    const active = detail.versions?.filter((v) => !v.is_yanked) ?? [];
    const latest = active[0]?.version;
    if (!latest) {
      throw new Error(`No installable versions for skill "${name}".`);
    }
    version = latest;
  }
  return fetchJson<VersionDetail>(
    `/skills/${encodeURIComponent(name)}/versions/${encodeURIComponent(version)}`,
  );
}
